### ISSUE TITLE
AST factory function call failing!?

### ISSUE THREAD

**Issue by erfajo** [DynamoDS/Dynamo#9512]
## Dynamo version
2.0.2  --> .Net 4.7

## Operating system
Windows 10, Revit 2019.1

## What did you do? 
I am building a dropdown node for IFCExportConfiguration (based on the IFC exporter API), and the code is as following...

```
public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
{
    if (Items.Count == 0 || Items[0].Name == noItemMessage || SelectedIndex == -1)
    { return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), AstFactory.BuildNullNode()) }; }

    var args = new List<AssociativeNode>
    { AstFactory.BuildStringNode(((IFCExportConfiguration) Items[SelectedIndex].Item).Name) };

    //var func = new Func<string, object>(GetConfiguration);
    //var assNode = AstFactory.BuildFunctionCall(func, args);

    var assNode = AstFactory.BuildFunctionCall
        <string, IFCExportConfiguration>(GetConfiguration, args);

    //var name = ((IFCExportConfiguration)Items[SelectedIndex].Item).Name;
    //var assNode = AstFactory.BuildPrimitiveNodeFromObject(GetConfiguration(name));

    return new[] { AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), assNode) };
}
```

## What did you expect to see?
I expected to get the selected IFCExportConfiguration 

## What did you see instead?
Even I can verify that my function "GetConfiguration(name)" return the IFCExportConfiguration, the AstFactory.BuildFunctionCall fail to handle it!?

![image](https://user-images.githubusercontent.com/2844374/53298004-35436500-3827-11e9-8b76-af585103ac0d.png)

I have tried all possible variations (see the commented lines) of generating the return object. I have even trying to cast it as an object.

**Comment by mjkkirschner** [comment]
make sure your function is defined in another assembly. It cannot be defined the same assembly as the nodeModel.

**Comment by erfajo** [comment]
I tried also to have my fynctions in another assembly, I found it somewhere that this could be an isssue... but it didnt help

```
public static List<IFCExportConfiguration> Configurations
{
    get
    {
        var configurations = new IFCExportConfigurationsMap();
        configurations.Add(IFCExportConfiguration.GetInSession());
        configurations.AddBuiltInConfigurations();
        configurations.AddSavedConfigurations();
        return configurations.Values.ToList();
    }
}

public static IFCExportConfiguration GetConfiguration(string name)
{
    IFCExportConfiguration result = null;
    foreach (IFCExportConfiguration configuration in Configurations)
    {
        if (name == configuration.Name)
        {
            result = configuration;
            break;
        }
    }
    return result;
}
```

**Comment by mjkkirschner** [comment]
when you use this node what result do you actually get as a result?
Have you debugged? - Does your function ever get called?

**Comment by erfajo** [comment]
I am debugging... but I can catch the function, thats why I made this topic.
If I use this function which fails, it do call my function
```
var name = ((IFCExportConfiguration)Items[SelectedIndex].Item).Name;
var assNode = AstFactory.BuildPrimitiveNodeFromObject(Orchid.Core.IFC.GetConfiguration(name));
```
Thats why I revealed my code, I cant figure what I am doing wrong.

**Comment by mjkkirschner** [comment]
`buildPrimitive` will only work on primitives - int,strings etc.
I'm sure you are aware, but just want to be 100% clear that the function you call must be in another assmebly (dll) not just another namespace to be called by the AST as a functionCall. So I would expect you to call it like 

```
 var functionCallNode = AstFactory.BuildFunctionCall
        <string, IFCExportConfiguration>(Orchid.Core.IFC.GetConfiguration, args);
```

**Comment by erfajo** [comment]
> `buildPrimitive` will only work on primitives - int,strings etc.
I know... but I had to do something to see that it could be called :-)

Orchid.dll holds Orchid.Core.IFC.GetConfiguration and TestBedUI.dll holds TestBed.UI.IFC.SettingDD
-and when it is being fixed will I move it from TestBedUI.dll to OrchidRevitUI.dll

**Comment by mjkkirschner** [comment]
so with the latest code you have above, there is no exception but your function never gets called?
It's a bit tough to guess what other setup could be causing trouble without having a repro example -I would ask that you produce the smallest example repo where you find the same issue and we can investigate that find missing documentation or a bug.

If thats not an option I would suggest
1. trying a simpler test function
2. making sure the node is being imported correctly (do you see the correct UI on it?)
3 . not importing any assemblies twice? Check the mange node and package paths
4. have some other package installed that might have some conflict, maybe an older version of your package?
5. do other dropdowns work? - like the dropdown in the dynamoSamples repo?

**Comment by erfajo** [comment]
Here is the bare minimum. I have compiled it and tested that It can make the needed node. However, still failing to run. I have included a minim graph as well with the node inserted.
[Demo.zip.txt](https://github.com/DynamoDS/Dynamo/files/2909945/Demo.zip.txt)

Remember besides the references for a normal Dynamo ZT project, you also need the assembly from the IFC esporter project
https://github.com/Autodesk/revit-ifc
